### ISSUE TITLE
sclang crash defining a SynthDef name with more than 255 characters

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#2803]
Hate to distract from more important work, but this seems easy to fix.

    SynthDef(($a!256).join, {}).add; // crash!

SynthDef names can only be up to 255 characters long due to the Pascal-style string format, so this should throw an error.

**Comment by brianlheim** [comment]
Easy fix: make `putPascalString` throw an error if its string argument has size greater than 255.

Hard (better) fix: use a different format (C-string) for writing/reading the name.

The call tree is
```supercollider
SynthDef:-add
  SynthDef:-asSynthDesc
    SynthDef:-asBytes
      Collection:-writeDef
        SynthDef:-writeDef
          CollStream:-putPascalString // should error here
  SynthDescLib:-readDescFromDef
    SynthDescLib:-readSynthDef2
      CollStream:-getPascalString
      FloatArray:*newClear // crash happens here when (for me) input is >= 536870897, pretty likley with garbage data
```

So this is really two issues in one—`putPascalString` doesn't do proper input validation, and `FloatArray.newClear` also crashes if its input is large enough.

**Comment by snappizz** [comment]
> Hard (better) fix: use a different format (C-string) for writing/reading the name.

We shouldn't change the SynthDef format and break alternate clients, and existing SynthDefs on hard drives, without a good reason.

Throwing an error is best. Nobody needs a SynthDef with a name that long.

**Comment by brianlheim** [comment]
> We shouldn't change the SynthDef format and break alternate clients, and existing SynthDefs on hard drives, without a good reason.

There's already a versioning system in place; it would not be a breaking (nor difficult) change to introduce a new format that uses C-strings and is read in with `readSynthDef3`.

> Nobody needs a SynthDef with a name that long.

Until somebody does.... Good implementation shouldn't needlessly restrict usage.

**Comment by muellmusik** [comment]
> On 20 Mar 2017, at 15:29, Brian Heim <notifications@github.com> wrote:
> 
> We shouldn't change the SynthDef format and break alternate clients, and existing SynthDefs on hard drives, without a good reason.
> 
> There's already a versioning system in place; it would not be a breaking (nor difficult) change to introduce a new format that uses C-strings and is read in with readSynthDef3.
> 
> 
Sorry, I’m with Nathan on this. It’s introducing a change for no real additional gain. Yes, you can introduce a new version, but that is not costless and would cause compatibility issues.
> Nobody needs a SynthDef with a name that long.
> 
> Until somebody does.... Good implementation shouldn't needlessly restrict usage.
> 
> 
Sure, but nor should it impose costs on others for a feature nobody wants at the moment, even if one day they might. I’d suggest waiting until there’s at least a tangible use case.

**Comment by jamshark70** [comment]
+1 for Scott and Nathan. To cause a concrete backward compatibility issue for non-sclang clients now, for an issue that has been in SC since its initial release 14-15 years ago and which has concretely affected exactly zero users (well, perhaps one) in that entire time... no. We have real problems to solve. Don't waste time on this.

**Comment by brianlheim** [comment]
> We have real problems to solve. Don't waste time on this.

Choose your words more kindly, please.

The 'real problems' we have to solve in SC are overwhelmingly generated by poor design decisions like `putPascalString` and poor patches for those poor design decisions. I really don't care which solution we go with in this case, because I think they're both acceptable. I just prefer playing devil's advocate for the more robust solution because I see a trend in this project of opting for the quick fix.

**Comment by muellmusik** [comment]
> Choose your words more kindly, please.

Well, one could make a similar request in response to critiques of 'poor design decisions', some of which might have been made 10 or 20 years ago in a rather different computing environment. There's an awful lot of judgement that flies around about stuff like that, and none of it helps make SC better.

But the main point is that we feel the cost is not worth the benefit. If you really want very long synthdef names (though I'm guessing it's not really that important for you?) then I think it would be more practical to make the change whenever a new synthdef version with other changes is ready to go through.

**Comment by vivid-synth** [comment]
Good eye, @snappizz . Looks from a quick glance we have the same problem with all other pascal strings (synthdef parameter names, ugen names, and variant names).

**Comment by vivid-synth** [comment]
I agree we shouldn't switch to C-style strings now, but I think @brianlheim 's instinct for an eye to the future is a good one. My (long-term, low-ish priority) proposal is to:

  - define an official v3 synthdef format which tidies up historical issues
  - write our clients to support (but not use) the new synthdef format
  - wait for a couple years to elapse
  - announce a switchover, at which point all clients will be expected to have gotten their SDv3 houses in order

**Comment by brianlheim** [comment]
> Well, one could make a similar request in response to critiques of 'poor design decisions', some of which might have been made 10 or 20 years ago in a rather different computing environment. There's an awful lot of judgement that flies around about stuff like that, and none of it helps make SC better.

I'm fully aware that a decision or good solution isn't always available for one reason or another, whether it's technology or time or other reasons. I don't judge anyone for that. I would like to think that by pointing it out and discussing we can stop repeating or building on old mistakes (I've seen at least a couple PRs in the last month or so that IMO do just that because there's not enough time or interest to fix the underlying issue), and ultimately make the codebase more maintainable. Also I just want to add that I'm speaking frankly because I'm speaking generally, and I'm sorry that I didn't do a good job of expressing that in a productive way. I think any reasonable developer would acknowledge that there are instances of poor design in this codebase, but I agree that directly accusing people or groups of people would not help make SC or its community better.

> I think it would be more practical to make the change whenever a new synthdef version with other changes is ready to go through.

Totally agree.

**Comment by jamshark70** [comment]
I apologize for the harsh tone.

To put it in a little context: Yesterday I found that any user on current master who drags a GUI object across an area of the screen governed by a container View (CompositeView, historically) will now get dozens of errors filling the post window. https://github.com/supercollider/supercollider/issues/2804 That's a **user-facing issue** (a "real problem") that so far has not earned any other developer's attention... while an issue that bothered no one in over 14 years inspired suggestions to break compatibility. I found those priorities difficult to understand.

The other part of the heat in my response is that we've also had a habit in this project of proposing changes that benefit the sclang client, but force authors of other clients to adapt, without any warning or appearance of consideration for those other clients. I can agree that this is a case of a poor design decision putting some limit on the user. That the limit is so painful for the user to justify making Hanns (scala client) upset is... dubious at best.

I have no objection to changing the string format if it is accompanied by other functional changes that carry a concrete benefit for all scsynth environments, not only sclang ones.
